### PR TITLE
Login + Playlist sync for Homey Media

### PR DESCRIPTION
**Pull request by MarvinSchenkel**
Adds login and playlist sync to the youtube app.

Note, nearly processed all your feedback. Don't have that much time, so maybe you can pick up a few things?

### CODE REVIEW

File: lib/youtube-mp3.js [hunk 1]
```diff
@@ -1,185 +1,26 @@
...
-    }).catch(callback)
+  var path = '/fetch/'
+
+  http.json({
+    protocol: 'http:',
+    host: mp3Url,
+    path: path,
+    query: {
+      format: 'JSON',
+      video: escape(videoUrl)
```
Comment (matjaz) [inline lib/youtube-mp3.js]: you shouldn't escape values here.

File: locales/nl.json [hunk 2]
```diff
@@ -4,6 +4,12 @@
 		"playlist": "Afspeelijst",
 		"playlistDelay": "Vertraging volgende",
 		"seconds": "Seconden",
-		"save": "Instellingen opslaan"
+		"save": "Instellingen opslaan",
+    "apiKey": "API key",
```
Comment (matjaz) [inline locales/nl.json]: please use same indentation.

File: lib/youtube-mp3.js [hunk 3]
```diff
@@ -1,185 +1,26 @@
...
+    }
+  }).then(function (response) {
+    if (!response.link) {
+      callback(new Error('Failed to get an MP3-stream. This is most likely because the video is longer than 120 minutes'))
```
Comment (matjaz) [inline lib/youtube-mp3.js]: missing return here

File: lib/oauth.js [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+      Homey.manager('api').realtime('authorized', false)
+      return Homey.error(err)
+    })
+  }
+}
+
+module.exports.getAccessToken = function (callback) {
+  var auth = Homey.manager('settings').get('auth')
+  if (!auth || !auth.accessToken || auth.accessToken === '') {
+    callback(new Error('Could not find an access token for the current user'))
```
Comment (matjaz) [inline lib/oauth.js:77]: missing return

File: lib/media.js [hunk 5]
```diff
@@ -123,8 +193,62 @@ exports.parseTrack = function (searchResult) {
...
+  })
+}
+
+function startPollingForUpdates () {
+  console.log('Polling for playlists..')
+  pollingInterval = setInterval(() => {
+    // Check if the user has logged in
+    oauth.getAccessToken(function (err, accessToken) {
+      if (err) {
+        return Homey.error(err)
```
Comment (matjaz) [inline lib/media.js]: return not needed

File: lib/youtube-mp3.js [hunk 6]
```diff
@@ -1,185 +1,26 @@
...
+    }
+    callback(null, response.link, response.length)
```
Comment (matjaz) [inline lib/youtube-mp3.js]: please convert this to `callback(err, val)` so
```
callback(null, {link, length})
```

File: lib/media.js [hunk 7]
```diff
@@ -82,30 +153,29 @@ exports.playVideoId = function (track, callback) {
...
-      duration = util.parsePeriod(item.contentDetails.duration) * 1000
-      // if we have a negative or 0 duration omit it
-      if (duration < 1) {
-        duration = undefined
-      }
-    }
     var track = exports.parseTrack(item)
     track.stream_url = url
-    track.duration = duration
+    // track.duration = duration
```
Comment (matjaz) [inline lib/media.js]: remove if not needed

File: lib/media.js [hunk 8]
```diff
@@ -25,16 +32,81 @@ exports.init = function () {
...
+        protocol: 'https:',
+        host: 'www.googleapis.com',
+        path: '/youtube/v3/playlists?part=id,snippet&mine=true&maxResults=' + MAX_PLAYLISTS + '&access_token=' + accessToken
+      }).then(function (response) {
+        if (response && response.items) {
+          var playlistRequests = response.items.map((item) => {
+            return new Promise((resolve) => {
+              getPlaylistItems(item.id, function (err, playlistItems) {
+                if (err) {
+                  callback(err, [])
```
Comment (matjaz) [inline lib/media.js]: we should reject promise or `return` here. I recommend reject and handle rejection in `Promise.all` below.

File: lib/media.js [hunk 9]
```diff
@@ -123,8 +193,62 @@ exports.parseTrack = function (searchResult) {
...
+
+      // Get details for all the tracks in a playlist
+      youtube('videos', {
+        part: 'id,snippet,contentDetails',
+        id: playlistIdList
+      }, function (err, result) {
+        if (err) {
+          callback(err)
+        } else {
+          var playlistItems = result.items.map(function (item) {
```
Comment (matjaz) [inline lib/media.js]: `var playlistItems = result.items.map(exports.parseTrack)`

**Review by matjaz — CHANGES_REQUESTED** [review]

**Comment by MarvinSchenkel** [comment]
Implemented your review changes!

**Review by matjaz — APPROVED** [review]

**Pull request merged**